src/entities/causalChain.ts

```diff
@@ -0,0 +1,263 @@
+     * Stage an event for integration into the chain.
+     * Returns whether or not the event can be integrated.
+     * @param {T} event
+     * @param {Immutable.Set<ModelComponent>} before
+     * @param {Immutable.Set<ModelComponent>} after
+     * @returns {boolean}
+     */
+    public stage(event: T,
+                 before: Immutable.Set<S>,
+                 after: Immutable.Set<S>): boolean {
+
+        let periodic = false;
+        let newLink = new Link<S, T>(event, before, after);
+        // try to find a link to connect with from among the tails
+        this.tails.forEach(link => {
+            if (link) {
+                let success = link.linkTo(newLink);
+                if (success) {
+                    if (newLink.hasSimilarAncestor(newLink)) {
+                        newLink.severLink(link);
+                        periodic = true;
+                    }
+                }
+            }
+        });
+
+        // if the new link does not become part of the
+        // execution chain because it causes periodicity,
+        // then it shouldn't be staged
+        if (!periodic) {
+            this.stagingArea.add(newLink);
+        }
+
+        return periodic;
+    }
+
+    /**
+     * Fully integrate the staged events into the chain, thereby
+     * updating the tails and clearing the staging area for the
+     * next integration round.
+     */
+    public finalize(): void {
+        let forRemoval = new Collections.Set<Link<S, T>>();
+        this.tails.forEach(link => {
+            // there has been a break in the causality
+            // so this link should be removed from
+            // the tail end of the chain
+            if (link.incident.isEmpty()) {
+                forRemoval.add(link);
+            }
+        });
+
+        forRemoval.forEach(toBeRemoved => {
+            this.tails.remove(toBeRemoved);
+        });
+
+        this.stagingArea.forEach(link => {
+            // make sure that the newly added tail is in fact a tail
+            // i.e. remove the antecedent links of the new tail
+            // from the list of tails
+            link.antecedent.forEach(ancestor => {
+                this.tails.remove(ancestor);
+            });
+            this.tails.add(link);
+        });
+
+        this.stagingArea.clear();
+    }
+
+    /**
+     * Retrieve the events residing in the tail.
+     * @returns {Set<Link<S, T>>}
+     */
+    public getTails(): Collections.Set<Link<S, T>> {
+        return this.tails;
+    }
+
+    /**
+     * Retrieve a string representation of the tail of this chain.
+     * Ideal for visualising incremental progress.
+     * @returns {string}
+     */
+    public toString(): string {
+        let chain = "";
+        this.tails.forEach(link => {
+            chain += "[";
+            link.antecedent.forEach(ancestor => {
+                chain += "(" + ancestor.toString() + ", " + ancestor.getEvent() + ")";
+            });
+            chain += "] => (";
```

Comment:
You could make this a little more readable by using string interpolation. 

```ts
chain += `] => (${link}, ${link.getEvent()})\n`;
```

Reply:
Ditto for the line above.

Reply:
Did not know that was a thing, thanks :)